### PR TITLE
Add support for grpc tracing header.

### PR DESCRIPTION
**Pull request by bogdandrutu**
The encoding is defined here:
https://github.com/census-instrumentation/opencensus-specs/blob/master/encodings/BinaryEncoding.md#trace-context

### CODE REVIEW

**Comment by qiwzhang** [comment]
@bogdandrutu could you fix the code format

Checking coding styles.
Files not formatted: src/api_manager/cloud_trace/cloud_trace.cc src/api_manager/cloud_trace/cloud_trace.h src/api_manager/cloud_trace/cloud_trace_test.cc src/api_manager/context/request_context.cc
Code style check failed.

You can run script/robot-style to check it.  We are using clang-3.8.0

File: src/api_manager/cloud_trace/cloud_trace.h [hunk 1]
```diff
@@ -139,10 +145,16 @@ class CloudTrace final {
...
 
+  const HeaderType header_type() const { return header_type_; }
+
+  std::string ToTraceContextHeader(uint64_t span_id) const;
+
  private:
   std::unique_ptr<google::devtools::cloudtrace::v1::Trace> trace_;
   google::devtools::cloudtrace::v1::TraceSpan *root_span_;
   std::string options_;
+  std::string original_trace_context_;
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply src/api_manager/cloud_trace/cloud_trace.h]: Removed. Thanks.

File: src/api_manager/context/request_context.cc [hunk 2]
```diff
@@ -358,17 +368,25 @@ const std::string RequestContext::FindClientIPAddress() {
...
+    Status status = request()->AddHeaderToBackend(kCloudTraceContextHeader,
+                                                  trace_context);
+    if (!status.ok()) {
+      service_context()->env()->LogError(
+          "Failed to set trace context header to backend.");
+    }
+  } else {
+    // Set grpc trace context header to backend.
+    Status status =
+        request()->AddHeaderToBackend(kGRpcTraceContextHeader, trace_context);
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply src/api_manager/context/request_context.cc]: Done.

**Comment by bogdandrutu** [comment]
Made progress on being able to build and run clang-format in a docker image, I cannot run the nginx rests (I will spend some time on that).

I have some questions:
1. Do you mind taking a dependency on absl?
2. Can you change your own trace_client implementation with OpenCensus? https://github.com/census-instrumentation/opencensus-cpp

**Comment by qiwzhang** [comment]
@bogdandrutu 
* we dont mind you use absl.
* we may not have resource to change to use OpenCensus.  But if there is business need, we may change priority.

**Comment by eli-jordan** [comment]
@qiwzhang @bogdandrutu 
I just hit this issue, and noticed the change in then PR behaves in a bit of a strange way. When running ESP with a gRPC backend:
- If invoked directly over gRPC the `grpc-trace-bin` header is forwarded to the backend
- If invoked over HTTP (i.e. using the transcoding feature) the `x-cloud-trace-context` header is forwarded to the backend.

If the backend is gRPC, I would expect the `grpc-trace-bin` to always be sent.

**Comment by qiwzhang** [comment]
@bogdandrutu this is the feature:  auto detect backend protocol use grpc-trace-bin or x-cloud-trace-context properly.   I remember you were about to finish this task.  What is your plan?

**Comment by qiwzhang** [comment]
I used a separate issue to track it: https://github.com/cloudendpoints/esp/issues/627

**Pull request merged**